{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Tutorial: 10x multiome pbmc\n",
    "\n",
    "\n",
    "The data consists of *PBMC from a Healthy Donor - Granulocytes Removed Through Cell Sorting (3k)* which is freely available from 10x Genomics (click [here](https://www.10xgenomics.com/resources/datasets/pbmc-from-a-healthy-donor-granulocytes-removed-through-cell-sorting-3-k-1-standard-2-0-0), some personal information needs to be provided before you can gain access to the data). This is a multi-ome dataset.\n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "\n",
    "**Note:**\n",
    "\n",
    "In this notebook we will only show the minimal steps needed for running the SCENIC+ analysis. For more information on analysing scRNA-seq data and scATAC-seq data we refer the reader to other tutorials (e.g. [Scanpy](https://scanpy-tutorials.readthedocs.io/en/latest/index.html) and [pycisTopic](https://pycistopic.readthedocs.io/en/latest/) in python or [Seurat](https://satijalab.org/seurat/) and [cisTopic](https://github.com/aertslab/cisTopic) or [Signac](https://satijalab.org/signac/) in R).\n",
    "\n",
    "</div>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set-up environment and download data \n",
    "We will first create a directory to store the data and results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pycisTopic.loom import *\n",
    "import itertools\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "import dill\n",
    "import string\n",
    "import pickle\n",
    "work_dir = '/work/rsenovilla/scenic/'\n",
    "tmp_dir = '/work/rsenovilla/scenic/temp/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#supress warnings\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "import sys\n",
    "import os\n",
    "_stderr = sys.stderr                                                         \n",
    "null = open(os.devnull,'wb')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## scRNA-seq preprocessing using Scanpy\n",
    "First we preprocess the scRNA-seq side of the mutliome datasets. Most importantly we will use this side of the data to annotate celltypes. \n",
    "\n",
    "For this we will make use of [Scanpy](https://scanpy.readthedocs.io/en/stable/). \n",
    "\n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "\n",
    "**Note:**\n",
    "\n",
    "You may also use [Seurat](https://satijalab.org/seurat/) (or any other tool in fact) to preprocess your data, however this will require some extra steps to import the data in python.\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "\n",
    "**Note:**\n",
    "\n",
    "Further on in the actual SCENIC+ analysis the raw count matrix will be used.\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating a cisTopic object and topic modeling\n",
    "\n",
    "Now that we have good quality barcodes we will generate a binary count matrix of ATAC-seq fragments over consensus peaks. This matrix, along with metadata, will be stored in a cisTopic object and be used for topic modeling.\n",
    "\n",
    "We will start by reading cell metadata from the scRNA-seq side of the analysis. For SCENIC+ we will only keep cells which passed quality metrics in both assays.\n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "\n",
    "**Note:**\n",
    "\n",
    "For independent scATAC-seq analysis you probably want to keep all cells (not only the cells passing the scRNA-seq filters). \n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-07-25 17:24:54,349 cisTopic     INFO     Converting fragment matrix to sparse matrix\n",
      "2025-07-25 17:26:00,385 cisTopic     INFO     Creating CistopicObject\n",
      "2025-07-25 17:26:05,704 cisTopic     INFO     Done!\n"
     ]
    }
   ],
   "source": [
    "# Create cisTopic object\n",
    "from pycisTopic.cistopic_class import *\n",
    "count_matrix=pd.read_csv(work_dir+'data_ce3/bgalgal1_atac_counts_080825.tsv', sep = '\\t', index_col = 0)\n",
    "count_matrix.index = count_matrix.index.astype(str).str.replace(\"-\", \":\", n=1)\n",
    "#path_to_blacklist=work_dir+'data_ce3/mm10-blacklist.v2.bed.gz'\n",
    "cistopic_obj = create_cistopic_object(fragment_matrix=count_matrix) #, path_to_blacklist=path_to_blacklist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_data =  pd.read_csv(work_dir+'data_ce3/ce3_cell_data_080825.tsv', index_col=0, sep = \"\\t\")\n",
    "cell_data['barcode'] = cell_data.index\n",
    "cell_data['cell_type_orig'] = cell_data['cell_types']\n",
    "cell_data['cell_type'] = cell_data['cell_types']\n",
    "cell_data['barcode'] = cell_data['barcode']+'___cisTopic'\n",
    "cell_data.index = cell_data['barcode'].astype(str)\n",
    "cell_data = cell_data.rename_axis(None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cisTopic_nr_frag</th>\n",
       "      <th>cisTopic_log_nr_frag</th>\n",
       "      <th>cisTopic_nr_acc</th>\n",
       "      <th>cisTopic_log_nr_acc</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_ATAC</th>\n",
       "      <th>nFeature_ATAC</th>\n",
       "      <th>total</th>\n",
       "      <th>duplicate</th>\n",
       "      <th>...</th>\n",
       "      <th>prediction.score.dTel</th>\n",
       "      <th>prediction.score.Mes.FP</th>\n",
       "      <th>prediction.score.Rho.RP</th>\n",
       "      <th>prediction.score.DLX.N</th>\n",
       "      <th>prediction.score.TBX20.N</th>\n",
       "      <th>prediction.score.Rho.vFP</th>\n",
       "      <th>cell_types_20052025</th>\n",
       "      <th>barcode</th>\n",
       "      <th>cell_type_orig</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ce3_1_AAACGAAAGAATACTG-1___cisTopic</th>\n",
       "      <td>4519</td>\n",
       "      <td>3.655042</td>\n",
       "      <td>4205</td>\n",
       "      <td>3.623766</td>\n",
       "      <td>cisTopic</td>\n",
       "      <td>ce3_1</td>\n",
       "      <td>4519</td>\n",
       "      <td>4205</td>\n",
       "      <td>8907</td>\n",
       "      <td>918</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.043422</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.049408</td>\n",
       "      <td>HB-FP</td>\n",
       "      <td>ce3_1_AAACGAAAGAATACTG-1___cisTopic</td>\n",
       "      <td>Rho-dFP</td>\n",
       "      <td>Rho-dFP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ce3_1_AAACGAAAGACCCTAT-1___cisTopic</th>\n",
       "      <td>5406</td>\n",
       "      <td>3.732876</td>\n",
       "      <td>5009</td>\n",
       "      <td>3.699751</td>\n",
       "      <td>cisTopic</td>\n",
       "      <td>ce3_1</td>\n",
       "      <td>5406</td>\n",
       "      <td>5009</td>\n",
       "      <td>9785</td>\n",
       "      <td>1025</td>\n",
       "      <td>...</td>\n",
       "      <td>0.024910</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OV</td>\n",
       "      <td>ce3_1_AAACGAAAGACCCTAT-1___cisTopic</td>\n",
       "      <td>OV</td>\n",
       "      <td>OV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ce3_1_AAACGAAAGACTAGGC-1___cisTopic</th>\n",
       "      <td>4743</td>\n",
       "      <td>3.676053</td>\n",
       "      <td>4398</td>\n",
       "      <td>3.643255</td>\n",
       "      <td>cisTopic</td>\n",
       "      <td>ce3_1</td>\n",
       "      <td>4743</td>\n",
       "      <td>4398</td>\n",
       "      <td>9022</td>\n",
       "      <td>833</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002223</td>\n",
       "      <td>0.017477</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>bP2</td>\n",
       "      <td>ce3_1_AAACGAAAGACTAGGC-1___cisTopic</td>\n",
       "      <td>bP2</td>\n",
       "      <td>bP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ce3_1_AAACGAAAGATGCGAC-1___cisTopic</th>\n",
       "      <td>7464</td>\n",
       "      <td>3.872972</td>\n",
       "      <td>6750</td>\n",
       "      <td>3.829304</td>\n",
       "      <td>cisTopic</td>\n",
       "      <td>ce3_1</td>\n",
       "      <td>7464</td>\n",
       "      <td>6750</td>\n",
       "      <td>14593</td>\n",
       "      <td>1116</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>aR1-2</td>\n",
       "      <td>ce3_1_AAACGAAAGATGCGAC-1___cisTopic</td>\n",
       "      <td>lR1-8</td>\n",
       "      <td>lR1-8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ce3_1_AAACGAAAGATGCGCA-1___cisTopic</th>\n",
       "      <td>6913</td>\n",
       "      <td>3.839667</td>\n",
       "      <td>6174</td>\n",
       "      <td>3.790567</td>\n",
       "      <td>cisTopic</td>\n",
       "      <td>ce3_1</td>\n",
       "      <td>6913</td>\n",
       "      <td>6174</td>\n",
       "      <td>12844</td>\n",
       "      <td>1187</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>P1</td>\n",
       "      <td>ce3_1_AAACGAAAGATGCGCA-1___cisTopic</td>\n",
       "      <td>bP2</td>\n",
       "      <td>bP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ce3_2_TTTGTGTGTAGTACAA-1___cisTopic</th>\n",
       "      <td>4718</td>\n",
       "      <td>3.673758</td>\n",
       "      <td>4362</td>\n",
       "      <td>3.639686</td>\n",
       "      <td>cisTopic</td>\n",
       "      <td>ce3_2</td>\n",
       "      <td>4718</td>\n",
       "      <td>4362</td>\n",
       "      <td>10230</td>\n",
       "      <td>932</td>\n",
       "      <td>...</td>\n",
       "      <td>0.278440</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>aPal</td>\n",
       "      <td>ce3_2_TTTGTGTGTAGTACAA-1___cisTopic</td>\n",
       "      <td>AMP</td>\n",
       "      <td>AMP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ce3_2_TTTGTGTGTCACTCTC-1___cisTopic</th>\n",
       "      <td>3408</td>\n",
       "      <td>3.5325</td>\n",
       "      <td>3251</td>\n",
       "      <td>3.512017</td>\n",
       "      <td>cisTopic</td>\n",
       "      <td>ce3_2</td>\n",
       "      <td>3408</td>\n",
       "      <td>3251</td>\n",
       "      <td>7385</td>\n",
       "      <td>475</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>aIT</td>\n",
       "      <td>ce3_2_TTTGTGTGTCACTCTC-1___cisTopic</td>\n",
       "      <td>a'It</td>\n",
       "      <td>a'It</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ce3_2_TTTGTGTGTTTGTAGC-1___cisTopic</th>\n",
       "      <td>4838</td>\n",
       "      <td>3.684666</td>\n",
       "      <td>4580</td>\n",
       "      <td>3.660865</td>\n",
       "      <td>cisTopic</td>\n",
       "      <td>ce3_2</td>\n",
       "      <td>4838</td>\n",
       "      <td>4580</td>\n",
       "      <td>10969</td>\n",
       "      <td>863</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>RL</td>\n",
       "      <td>ce3_2_TTTGTGTGTTTGTAGC-1___cisTopic</td>\n",
       "      <td>LHX9-N</td>\n",
       "      <td>LHX9-N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ce3_2_TTTGTGTTCAAAGTAG-1___cisTopic</th>\n",
       "      <td>3651</td>\n",
       "      <td>3.562412</td>\n",
       "      <td>3463</td>\n",
       "      <td>3.539452</td>\n",
       "      <td>cisTopic</td>\n",
       "      <td>ce3_2</td>\n",
       "      <td>3651</td>\n",
       "      <td>3463</td>\n",
       "      <td>7747</td>\n",
       "      <td>761</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.701748</td>\n",
       "      <td>HB-FP</td>\n",
       "      <td>ce3_2_TTTGTGTTCAAAGTAG-1___cisTopic</td>\n",
       "      <td>Rho-vFP</td>\n",
       "      <td>Rho-vFP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ce3_2_TTTGTGTTCATCATGG-1___cisTopic</th>\n",
       "      <td>4085</td>\n",
       "      <td>3.611192</td>\n",
       "      <td>3872</td>\n",
       "      <td>3.587935</td>\n",
       "      <td>cisTopic</td>\n",
       "      <td>ce3_2</td>\n",
       "      <td>4085</td>\n",
       "      <td>3872</td>\n",
       "      <td>8832</td>\n",
       "      <td>848</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013228</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.686299</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>MOT N</td>\n",
       "      <td>ce3_2_TTTGTGTTCATCATGG-1___cisTopic</td>\n",
       "      <td>TBX20-N</td>\n",
       "      <td>TBX20-N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28719 rows Ã— 123 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    cisTopic_nr_frag cisTopic_log_nr_frag  \\\n",
       "ce3_1_AAACGAAAGAATACTG-1___cisTopic             4519             3.655042   \n",
       "ce3_1_AAACGAAAGACCCTAT-1___cisTopic             5406             3.732876   \n",
       "ce3_1_AAACGAAAGACTAGGC-1___cisTopic             4743             3.676053   \n",
       "ce3_1_AAACGAAAGATGCGAC-1___cisTopic             7464             3.872972   \n",
       "ce3_1_AAACGAAAGATGCGCA-1___cisTopic             6913             3.839667   \n",
       "...                                              ...                  ...   \n",
       "ce3_2_TTTGTGTGTAGTACAA-1___cisTopic             4718             3.673758   \n",
       "ce3_2_TTTGTGTGTCACTCTC-1___cisTopic             3408               3.5325   \n",
       "ce3_2_TTTGTGTGTTTGTAGC-1___cisTopic             4838             3.684666   \n",
       "ce3_2_TTTGTGTTCAAAGTAG-1___cisTopic             3651             3.562412   \n",
       "ce3_2_TTTGTGTTCATCATGG-1___cisTopic             4085             3.611192   \n",
       "\n",
       "                                    cisTopic_nr_acc cisTopic_log_nr_acc  \\\n",
       "ce3_1_AAACGAAAGAATACTG-1___cisTopic            4205            3.623766   \n",
       "ce3_1_AAACGAAAGACCCTAT-1___cisTopic            5009            3.699751   \n",
       "ce3_1_AAACGAAAGACTAGGC-1___cisTopic            4398            3.643255   \n",
       "ce3_1_AAACGAAAGATGCGAC-1___cisTopic            6750            3.829304   \n",
       "ce3_1_AAACGAAAGATGCGCA-1___cisTopic            6174            3.790567   \n",
       "...                                             ...                 ...   \n",
       "ce3_2_TTTGTGTGTAGTACAA-1___cisTopic            4362            3.639686   \n",
       "ce3_2_TTTGTGTGTCACTCTC-1___cisTopic            3251            3.512017   \n",
       "ce3_2_TTTGTGTGTTTGTAGC-1___cisTopic            4580            3.660865   \n",
       "ce3_2_TTTGTGTTCAAAGTAG-1___cisTopic            3463            3.539452   \n",
       "ce3_2_TTTGTGTTCATCATGG-1___cisTopic            3872            3.587935   \n",
       "\n",
       "                                    sample_id orig.ident  nCount_ATAC  \\\n",
       "ce3_1_AAACGAAAGAATACTG-1___cisTopic  cisTopic      ce3_1         4519   \n",
       "ce3_1_AAACGAAAGACCCTAT-1___cisTopic  cisTopic      ce3_1         5406   \n",
       "ce3_1_AAACGAAAGACTAGGC-1___cisTopic  cisTopic      ce3_1         4743   \n",
       "ce3_1_AAACGAAAGATGCGAC-1___cisTopic  cisTopic      ce3_1         7464   \n",
       "ce3_1_AAACGAAAGATGCGCA-1___cisTopic  cisTopic      ce3_1         6913   \n",
       "...                                       ...        ...          ...   \n",
       "ce3_2_TTTGTGTGTAGTACAA-1___cisTopic  cisTopic      ce3_2         4718   \n",
       "ce3_2_TTTGTGTGTCACTCTC-1___cisTopic  cisTopic      ce3_2         3408   \n",
       "ce3_2_TTTGTGTGTTTGTAGC-1___cisTopic  cisTopic      ce3_2         4838   \n",
       "ce3_2_TTTGTGTTCAAAGTAG-1___cisTopic  cisTopic      ce3_2         3651   \n",
       "ce3_2_TTTGTGTTCATCATGG-1___cisTopic  cisTopic      ce3_2         4085   \n",
       "\n",
       "                                     nFeature_ATAC  total  duplicate  ...  \\\n",
       "ce3_1_AAACGAAAGAATACTG-1___cisTopic           4205   8907        918  ...   \n",
       "ce3_1_AAACGAAAGACCCTAT-1___cisTopic           5009   9785       1025  ...   \n",
       "ce3_1_AAACGAAAGACTAGGC-1___cisTopic           4398   9022        833  ...   \n",
       "ce3_1_AAACGAAAGATGCGAC-1___cisTopic           6750  14593       1116  ...   \n",
       "ce3_1_AAACGAAAGATGCGCA-1___cisTopic           6174  12844       1187  ...   \n",
       "...                                            ...    ...        ...  ...   \n",
       "ce3_2_TTTGTGTGTAGTACAA-1___cisTopic           4362  10230        932  ...   \n",
       "ce3_2_TTTGTGTGTCACTCTC-1___cisTopic           3251   7385        475  ...   \n",
       "ce3_2_TTTGTGTGTTTGTAGC-1___cisTopic           4580  10969        863  ...   \n",
       "ce3_2_TTTGTGTTCAAAGTAG-1___cisTopic           3463   7747        761  ...   \n",
       "ce3_2_TTTGTGTTCATCATGG-1___cisTopic           3872   8832        848  ...   \n",
       "\n",
       "                                     prediction.score.dTel  \\\n",
       "ce3_1_AAACGAAAGAATACTG-1___cisTopic               0.000000   \n",
       "ce3_1_AAACGAAAGACCCTAT-1___cisTopic               0.024910   \n",
       "ce3_1_AAACGAAAGACTAGGC-1___cisTopic               0.002223   \n",
       "ce3_1_AAACGAAAGATGCGAC-1___cisTopic               0.000000   \n",
       "ce3_1_AAACGAAAGATGCGCA-1___cisTopic               0.000000   \n",
       "...                                                    ...   \n",
       "ce3_2_TTTGTGTGTAGTACAA-1___cisTopic               0.278440   \n",
       "ce3_2_TTTGTGTGTCACTCTC-1___cisTopic               0.000000   \n",
       "ce3_2_TTTGTGTGTTTGTAGC-1___cisTopic               0.000000   \n",
       "ce3_2_TTTGTGTTCAAAGTAG-1___cisTopic               0.000000   \n",
       "ce3_2_TTTGTGTTCATCATGG-1___cisTopic               0.000000   \n",
       "\n",
       "                                     prediction.score.Mes.FP  \\\n",
       "ce3_1_AAACGAAAGAATACTG-1___cisTopic                 0.043422   \n",
       "ce3_1_AAACGAAAGACCCTAT-1___cisTopic                 0.000000   \n",
       "ce3_1_AAACGAAAGACTAGGC-1___cisTopic                 0.017477   \n",
       "ce3_1_AAACGAAAGATGCGAC-1___cisTopic                 0.000000   \n",
       "ce3_1_AAACGAAAGATGCGCA-1___cisTopic                 0.000000   \n",
       "...                                                      ...   \n",
       "ce3_2_TTTGTGTGTAGTACAA-1___cisTopic                 0.000000   \n",
       "ce3_2_TTTGTGTGTCACTCTC-1___cisTopic                 0.000000   \n",
       "ce3_2_TTTGTGTGTTTGTAGC-1___cisTopic                 0.000000   \n",
       "ce3_2_TTTGTGTTCAAAGTAG-1___cisTopic                 0.000000   \n",
       "ce3_2_TTTGTGTTCATCATGG-1___cisTopic                 0.013228   \n",
       "\n",
       "                                     prediction.score.Rho.RP  \\\n",
       "ce3_1_AAACGAAAGAATACTG-1___cisTopic                      0.0   \n",
       "ce3_1_AAACGAAAGACCCTAT-1___cisTopic                      0.0   \n",
       "ce3_1_AAACGAAAGACTAGGC-1___cisTopic                      0.0   \n",
       "ce3_1_AAACGAAAGATGCGAC-1___cisTopic                      0.0   \n",
       "ce3_1_AAACGAAAGATGCGCA-1___cisTopic                      0.0   \n",
       "...                                                      ...   \n",
       "ce3_2_TTTGTGTGTAGTACAA-1___cisTopic                      0.0   \n",
       "ce3_2_TTTGTGTGTCACTCTC-1___cisTopic                      0.0   \n",
       "ce3_2_TTTGTGTGTTTGTAGC-1___cisTopic                      0.0   \n",
       "ce3_2_TTTGTGTTCAAAGTAG-1___cisTopic                      0.0   \n",
       "ce3_2_TTTGTGTTCATCATGG-1___cisTopic                      0.0   \n",
       "\n",
       "                                     prediction.score.DLX.N  \\\n",
       "ce3_1_AAACGAAAGAATACTG-1___cisTopic                     0.0   \n",
       "ce3_1_AAACGAAAGACCCTAT-1___cisTopic                     0.0   \n",
       "ce3_1_AAACGAAAGACTAGGC-1___cisTopic                     0.0   \n",
       "ce3_1_AAACGAAAGATGCGAC-1___cisTopic                     0.0   \n",
       "ce3_1_AAACGAAAGATGCGCA-1___cisTopic                     0.0   \n",
       "...                                                     ...   \n",
       "ce3_2_TTTGTGTGTAGTACAA-1___cisTopic                     0.0   \n",
       "ce3_2_TTTGTGTGTCACTCTC-1___cisTopic                     0.0   \n",
       "ce3_2_TTTGTGTGTTTGTAGC-1___cisTopic                     0.0   \n",
       "ce3_2_TTTGTGTTCAAAGTAG-1___cisTopic                     0.0   \n",
       "ce3_2_TTTGTGTTCATCATGG-1___cisTopic                     0.0   \n",
       "\n",
       "                                     prediction.score.TBX20.N  \\\n",
       "ce3_1_AAACGAAAGAATACTG-1___cisTopic                  0.000000   \n",
       "ce3_1_AAACGAAAGACCCTAT-1___cisTopic                  0.000000   \n",
       "ce3_1_AAACGAAAGACTAGGC-1___cisTopic                  0.000000   \n",
       "ce3_1_AAACGAAAGATGCGAC-1___cisTopic                  0.000000   \n",
       "ce3_1_AAACGAAAGATGCGCA-1___cisTopic                  0.000000   \n",
       "...                                                       ...   \n",
       "ce3_2_TTTGTGTGTAGTACAA-1___cisTopic                  0.000000   \n",
       "ce3_2_TTTGTGTGTCACTCTC-1___cisTopic                  0.000000   \n",
       "ce3_2_TTTGTGTGTTTGTAGC-1___cisTopic                  0.000000   \n",
       "ce3_2_TTTGTGTTCAAAGTAG-1___cisTopic                  0.000000   \n",
       "ce3_2_TTTGTGTTCATCATGG-1___cisTopic                  0.686299   \n",
       "\n",
       "                                     prediction.score.Rho.vFP  \\\n",
       "ce3_1_AAACGAAAGAATACTG-1___cisTopic                  0.049408   \n",
       "ce3_1_AAACGAAAGACCCTAT-1___cisTopic                  0.000000   \n",
       "ce3_1_AAACGAAAGACTAGGC-1___cisTopic                  0.000000   \n",
       "ce3_1_AAACGAAAGATGCGAC-1___cisTopic                  0.000000   \n",
       "ce3_1_AAACGAAAGATGCGCA-1___cisTopic                  0.000000   \n",
       "...                                                       ...   \n",
       "ce3_2_TTTGTGTGTAGTACAA-1___cisTopic                  0.000000   \n",
       "ce3_2_TTTGTGTGTCACTCTC-1___cisTopic                  0.000000   \n",
       "ce3_2_TTTGTGTGTTTGTAGC-1___cisTopic                  0.000000   \n",
       "ce3_2_TTTGTGTTCAAAGTAG-1___cisTopic                  0.701748   \n",
       "ce3_2_TTTGTGTTCATCATGG-1___cisTopic                  0.000000   \n",
       "\n",
       "                                     cell_types_20052025  \\\n",
       "ce3_1_AAACGAAAGAATACTG-1___cisTopic                HB-FP   \n",
       "ce3_1_AAACGAAAGACCCTAT-1___cisTopic                   OV   \n",
       "ce3_1_AAACGAAAGACTAGGC-1___cisTopic                  bP2   \n",
       "ce3_1_AAACGAAAGATGCGAC-1___cisTopic                aR1-2   \n",
       "ce3_1_AAACGAAAGATGCGCA-1___cisTopic                   P1   \n",
       "...                                                  ...   \n",
       "ce3_2_TTTGTGTGTAGTACAA-1___cisTopic                 aPal   \n",
       "ce3_2_TTTGTGTGTCACTCTC-1___cisTopic                  aIT   \n",
       "ce3_2_TTTGTGTGTTTGTAGC-1___cisTopic                   RL   \n",
       "ce3_2_TTTGTGTTCAAAGTAG-1___cisTopic                HB-FP   \n",
       "ce3_2_TTTGTGTTCATCATGG-1___cisTopic                MOT N   \n",
       "\n",
       "                                                                 barcode  \\\n",
       "ce3_1_AAACGAAAGAATACTG-1___cisTopic  ce3_1_AAACGAAAGAATACTG-1___cisTopic   \n",
       "ce3_1_AAACGAAAGACCCTAT-1___cisTopic  ce3_1_AAACGAAAGACCCTAT-1___cisTopic   \n",
       "ce3_1_AAACGAAAGACTAGGC-1___cisTopic  ce3_1_AAACGAAAGACTAGGC-1___cisTopic   \n",
       "ce3_1_AAACGAAAGATGCGAC-1___cisTopic  ce3_1_AAACGAAAGATGCGAC-1___cisTopic   \n",
       "ce3_1_AAACGAAAGATGCGCA-1___cisTopic  ce3_1_AAACGAAAGATGCGCA-1___cisTopic   \n",
       "...                                                                  ...   \n",
       "ce3_2_TTTGTGTGTAGTACAA-1___cisTopic  ce3_2_TTTGTGTGTAGTACAA-1___cisTopic   \n",
       "ce3_2_TTTGTGTGTCACTCTC-1___cisTopic  ce3_2_TTTGTGTGTCACTCTC-1___cisTopic   \n",
       "ce3_2_TTTGTGTGTTTGTAGC-1___cisTopic  ce3_2_TTTGTGTGTTTGTAGC-1___cisTopic   \n",
       "ce3_2_TTTGTGTTCAAAGTAG-1___cisTopic  ce3_2_TTTGTGTTCAAAGTAG-1___cisTopic   \n",
       "ce3_2_TTTGTGTTCATCATGG-1___cisTopic  ce3_2_TTTGTGTTCATCATGG-1___cisTopic   \n",
       "\n",
       "                                     cell_type_orig  cell_type  \n",
       "ce3_1_AAACGAAAGAATACTG-1___cisTopic         Rho-dFP    Rho-dFP  \n",
       "ce3_1_AAACGAAAGACCCTAT-1___cisTopic              OV         OV  \n",
       "ce3_1_AAACGAAAGACTAGGC-1___cisTopic             bP2        bP2  \n",
       "ce3_1_AAACGAAAGATGCGAC-1___cisTopic           lR1-8      lR1-8  \n",
       "ce3_1_AAACGAAAGATGCGCA-1___cisTopic             bP2        bP2  \n",
       "...                                             ...        ...  \n",
       "ce3_2_TTTGTGTGTAGTACAA-1___cisTopic             AMP        AMP  \n",
       "ce3_2_TTTGTGTGTCACTCTC-1___cisTopic            a'It       a'It  \n",
       "ce3_2_TTTGTGTGTTTGTAGC-1___cisTopic          LHX9-N     LHX9-N  \n",
       "ce3_2_TTTGTGTTCAAAGTAG-1___cisTopic         Rho-vFP    Rho-vFP  \n",
       "ce3_2_TTTGTGTTCATCATGG-1___cisTopic         TBX20-N    TBX20-N  \n",
       "\n",
       "[28719 rows x 123 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_data['cell_type'].value_counts()\n",
    "cistopic_obj.add_cell_data(cell_data)\n",
    "cistopic_obj.cell_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#it takes fucking ages\n",
    "import pickle\n",
    "file_path = '/work/rsenovilla/scenic/atac_ce3/cistopic_obj_080825.pkl'\n",
    "\n",
    "# Use the 'wb' mode to write the pickle file\n",
    "with open(file_path, 'wb') as file:\n",
    "    pickle.dump(cistopic_obj, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run topic modeling. The purpose of this is twofold:\n",
    "\n",
    "1. To find sets of co-accessible regions (topics), this will be used downstream as candidate enhancers (together with Differentially Accessible Regions (DARs)).\n",
    "2. To impute dropouts. \n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "\n",
    "**Note:**\n",
    "\n",
    "scATAC-seq data is *very* sparse. This is because of the nature of the technique. It profiles chromatin accessibility in single cells and each cell only has two copies (two alleles) of each genomic region, also the genome is realtively large (so the number of regions which can be measured is large). For this reason drop outs are a *real* problem, the chance of measuring a specific genomic region (out of *many* potential regions) which only has two copies per cell is relatively small. Compare this to scRNA-seq analysis, here the number of genes which can be measure is relatively small (compared to the number of genomic regions) and each cell has potentially hundres of copies of each transcript. To account for drop outs in the scATAC-seq assay imputation techniques are often used, in this case we make use of topic modeling, making use of the fact that the data often contains cells which are quite similar to each other but might have slightly different features measured. \n",
    "</div>\n",
    "\n",
    "\n",
    "Before running the topic modeling we are not sure what the best number of topics will be to explain the data. Analog to PCA where you also often don't know before hand what the best number of principle components is. For this reason we will generate models with increasing numbers of topics and after the fact choose the model with the optimal amount of topics. For demonstration purposes we will only try a few amount of models, you might want to explore a larger number of topics.\n",
    "\n",
    "<div class=\"alert alert-warning\">\n",
    "\n",
    "**Warning:**\n",
    "\n",
    "Topic modeling can be computationaly intense!\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-07-25 17:37:14,092\tINFO worker.py:1633 -- Started a local Ray instance. View the dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265 \u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(run_cgs_model pid=3272179)\u001b[0m 2025-07-25 17:37:19,522 cisTopic     INFO     Running model with 30 topics\n"
     ]
    }
   ],
   "source": [
    "#Serial LDA: The parallelization is done between models rather than within each model. Recommended for small-medium sized data sets in which several models with different number of topics are being tested. \n",
    "import pickle\n",
    "cistopic_obj = pickle.load(open(os.path.join(work_dir, 'atac_ce3/cistopic_obj_080825.pkl'), 'rb'))\n",
    "from pycisTopic.lda_models import *\n",
    "models=run_cgs_models(cistopic_obj,\n",
    "                    n_topics=[30,35,40,45,50],\n",
    "                    n_cpu=60,\n",
    "                    n_iter=500,\n",
    "                    random_state=555,\n",
    "                    alpha=50,\n",
    "                      \n",
    "                    alpha_by_topic=True,\n",
    "                    eta=0.1,\n",
    "                    eta_by_topic=False,\n",
    "                    save_path=None,\n",
    "                    _temp_dir = os.path.join(tmp_dir + 'ray_spill'))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "#Parallel LDA with MALLET: The parallelization is done within each model. Recommended for large data sets where a few models with different number of topics are being tested. If working in a cluster, we recommed to submit a job per model so they can run simultaneously\n",
    "import os\n",
    "path_to_mallet_binary='mallet'\n",
    "# Run models\n",
    "#models=run_cgs_models_mallet(path_to_mallet_binary,\n",
    "                    cistopic_obj,\n",
    "                    n_topics=[2,5,10,15,20,25,30,35,40,45,50],\n",
    "                    n_cpu=24,\n",
    "                    n_iter=500,\n",
    "                    random_state=555,\n",
    "                    alpha=50,\n",
    "                    alpha_by_topic=True,\n",
    "                    eta=0.1,\n",
    "                    eta_by_topic=False,\n",
    "                    tmp_path= os.path.join(tmp_dir + 'scratch'), #Use SCRATCH if many models or big data set\n",
    "                    save_path=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analyze models.\n",
    "\n",
    "We will make use of four quality metrics to select the model with the optimal amount of topics:\n",
    "1. [Arun *et al.* 2010](http://link.springer.com/10.1007/978-3-642-13657-3_43)\n",
    "2. [Cao & Juan *et al.* 2009](https://linkinghub.elsevier.com/retrieve/pii/S092523120800372X)\n",
    "3. [Mimno *et al.* 2011](http://dirichlet.net/pdf/mimno11optimizing.pdf)\n",
    "4. Log likelihood\n",
    "\n",
    "For more information on these metrics see publications (linked above) and the [read the docs](https://pycistopic.readthedocs.io/en/latest/) page of pycisTopic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(os.path.join(work_dir, 'atac_ce3/models')):\n",
    "    os.makedirs(os.path.join(work_dir, 'atac_ce3/models'))\n",
    "pickle.dump(models,\n",
    "            open(os.path.join(work_dir, 'atac_ce3/models/ce3_models_500_iter_LDA_080825.pkl'), 'wb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "models = pickle.load(open(os.path.join(work_dir, 'atac_ce3/models/ce3_models_500_iter_LDA_080825.pkl'), 'rb'))\n",
    "cistopic_obj = pickle.load(open(os.path.join(work_dir, 'atac_ce3/cistopic_obj_080825.pkl'), 'rb'))\n",
    "from pycisTopic.lda_models import *\n",
    "model = evaluate_models(models,\n",
    "                       select_model=None, \n",
    "                       return_model=True, \n",
    "                       metrics=['Arun_2010','Cao_Juan_2009', 'Minmo_2011', 'loglikelihood'],\n",
    "                       plot_metrics=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The metrics seem to stabelise with a model using 16 topics, so let's choose that model. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cistopic_obj.add_LDA_model(model)\n",
    "pickle.dump(cistopic_obj,\n",
    "          open(os.path.join(work_dir, 'atac_ce3/cistopic_obj_080825.pkl'), 'wb'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Visualization\n",
    "\n",
    "We can use the cell-topic probabilities to generate dimensionality reductions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycisTopic.clust_vis import *\n",
    "run_umap(cistopic_obj, target  = 'cell', scale=True)\n",
    "plot_metadata(cistopic_obj, reduction_name = 'UMAP', variables = ['cell_type'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also plot the cell-topic probabilities on the UMAP, to visualize their cell type specifiticy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'UMAP'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[14], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mplot_topic\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcistopic_obj\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreduction_name\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mUMAP\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnum_columns\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m4\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/work/rsenovilla/miniconda3/envs/scenic/lib/python3.8/site-packages/pycisTopic/clust_vis.py:643\u001b[0m, in \u001b[0;36mplot_topic\u001b[0;34m(cistopic_obj, reduction_name, target, cmap, dot_size, alpha, scale, selected_topics, selected_features, harmony, figsize, num_columns, save)\u001b[0m\n\u001b[1;32m    593\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mplot_topic\u001b[39m(\n\u001b[1;32m    594\u001b[0m     cistopic_obj: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCistopicObject\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    595\u001b[0m     reduction_name: \u001b[38;5;28mstr\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    606\u001b[0m     save: Optional[\u001b[38;5;28mstr\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m    607\u001b[0m ):\n\u001b[1;32m    608\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    609\u001b[0m \u001b[38;5;124;03m    Plot topic distributions into dimensionality reduction.\u001b[39;00m\n\u001b[1;32m    610\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    640\u001b[0m \u001b[38;5;124;03m            Path to save plot. Default: None.\u001b[39;00m\n\u001b[1;32m    641\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 643\u001b[0m     embedding \u001b[38;5;241m=\u001b[39m \u001b[43mcistopic_obj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mprojections\u001b[49m\u001b[43m[\u001b[49m\u001b[43mtarget\u001b[49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[43mreduction_name\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m    644\u001b[0m     model \u001b[38;5;241m=\u001b[39m cistopic_obj\u001b[38;5;241m.\u001b[39mselected_model\n\u001b[1;32m    646\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m target \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcell\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n",
      "\u001b[0;31mKeyError\u001b[0m: 'UMAP'"
     ]
    }
   ],
   "source": [
    "plot_topic(cistopic_obj, reduction_name = 'UMAP', num_columns = 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For further analysis see the [pycisTopic read the docs page](https://pycistopic.readthedocs.io/en/latest/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inferring candidate enhancer regions\n",
    "\n",
    "Next we will infer candidate enhancer regions by:\n",
    "\n",
    "1. binarization of region-topic probabilites.\n",
    "2. calculation differentially accessibile regions (DARs) per cell type.\n",
    "\n",
    "These regions will be used as input for the next step, [pycistarget](https://pycistarget.readthedocs.io/en/latest/), in which we will look which motifs are enriched in these regions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First we will binarize the topics using the [otsu](http://ieeexplore.ieee.org/document/4310076/) method and by taking the top 3k regions per topic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycisTopic.topic_binarization import *\n",
    "region_bin_topics_otsu = binarize_topics(cistopic_obj, method='otsu')\n",
    "region_bin_topics_top3k = binarize_topics(cistopic_obj, method='ntop', ntop = 3000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we will calculate DARs per cell type"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "from pycisTopic.diff_features import *\n",
    "imputed_acc_obj = impute_accessibility(cistopic_obj, selected_cells=None, selected_regions=None, scale_factor=10**6)\n",
    "normalized_imputed_acc_obj = normalize_scores(imputed_acc_obj, scale_factor=10**4)\n",
    "variable_regions = find_highly_variable_features(normalized_imputed_acc_obj, plot = False)\n",
    "markers_dict = find_diff_features(cistopic_obj, imputed_acc_obj, variable='cell_type', var_features=variable_regions, split_pattern = '-')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cistopic_obj = pickle.load(open(os.path.join(work_dir, 'atac_ce3/cistopic_obj_080825.pkl'), 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#if args.mode == \"impute\" or args.mode==\"impute/non-impute\": \n",
    "from pycisTopic.diff_features import *\n",
    "path_to_imputed = \"/work/rsenovilla/scenic/atac_ce3/cistopic_imputed_obj_080825.pkl\"\n",
    "imputed_obj = impute_accessibility(\n",
    "        cistopic_obj, \n",
    "        selected_cells=None, \n",
    "        selected_regions=None, \n",
    "        scale_factor=10**6, \n",
    "        #sparsity_thr=0.67, #It does not appear as a parameter\n",
    "        project=\"cistopic_imputed\"\n",
    "    )\n",
    "with open(path_to_imputed, \"wb\") as f:\n",
    "        dill.dump(imputed_obj, f, protocol=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycisTopic.diff_features import *\n",
    "markers_dict = find_diff_features(\n",
    "        cistopic_obj, \n",
    "        imputed_obj,\n",
    "        variable='cell_type',\n",
    "        var_features=None,\n",
    "        contrasts=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#if args.mode == \"non-impute\" or args.mode==\"impute/non-impute\":\n",
    "path_to_non_imputed = \"/work/rsenovilla/scenic/atac_ce3/cistopic_nonimputed_obj_080825.pkl\"\n",
    "nonimputed_obj = CistopicImputedFeatures(\n",
    "        imputed_acc=cistopic_obj.fragment_matrix,\n",
    "        cell_names=cistopic_obj.cell_names,\n",
    "        feature_names=cistopic_obj.region_names,\n",
    "        project=\"cistopic_not_imputed\"\n",
    "    )\n",
    "    # remove zero rows as done by the impute_accessibility function\n",
    "zero_rows = (nonimputed_obj.mtx.sum(axis=1) == 0)\n",
    "nonimputed_obj.mtx = nonimputed_obj.mtx[~np.array(zero_rows).flatten(), :]\n",
    "nonimputed_obj.feature_names = list(np.array(nonimputed_obj.feature_names)[~np.array(zero_rows).flatten()])\n",
    "with open(path_to_non_imputed, \"wb\") as f:\n",
    "        dill.dump(nonimputed_obj, f, protocol=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(os.path.join(work_dir, 'atac_ce3/candidate_enhancers')):\n",
    "    os.makedirs(os.path.join(work_dir, 'atac_ce3/candidate_enhancers'))\n",
    "import pickle\n",
    "pickle.dump(region_bin_topics_otsu, open(os.path.join(work_dir, 'atac_ce3/candidate_enhancers/region_bin_topics_otsu_080825.pkl'), 'wb'))\n",
    "pickle.dump(region_bin_topics_top3k, open(os.path.join(work_dir, 'atac_ce3/candidate_enhancers/region_bin_topics_top3k_080825.pkl'), 'wb'))\n",
    "pickle.dump(markers_dict, open(os.path.join(work_dir, 'atac_ce3/candidate_enhancers/markers_dict_080825.pkl'), 'wb'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now completed all the mininal scATAC-seq preprocessing steps. \n",
    "\n",
    "In particular we:\n",
    "\n",
    "1. generated a set of consensus peaks\n",
    "2. performed quality control steps, only keeping cell barcods which passed QC metrics in both the scRNA-seq and scATAC-seq assay\n",
    "3. performed topic modeling\n",
    "4. inferred candidate enhancer regions by binarizing the region-topic probabilities and DARs per cell type\n",
    "\n",
    "In the next step we will perform motif enrichment analysis on these candidate enhancer regions using the python package, [pycistarget](phttps://pycistarget.readthedocs.io/en/latest/). For this a precomputed motif-score database is needed. A sample specific database can be generated by scoring the consensus peaks with motifs or a general pre-scored database can be used as well."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Motif enrichment analysis using pycistarget\n",
    "\n",
    "After having identified candidate enhancer regions we will use [pycistarget](https://pycistarget.readthedocs.io/en/latest/) to find which motifs are enriched in these regions. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cistarget databases\n",
    "\n",
    "In order to run pycistarget one needs a precomputed database containing motif scores for genomic regions.\n",
    "\n",
    "You can choose to compute this database yourself by scoring the consensus peaks generated in the scATAC-seq analysis using a set of motifs. The advantage of creating a sample specific database is that you can potentially pick up more target regions, given that only regions included/overlappig with regions in the cistarget database will be used for the SCENIC+ analysis. For more information checkout the [create_cisTarget_databases repo on github](https://github.com/aertslab/create_cisTarget_databases). \n",
    "\n",
    "We also provide several precomputed databases containing regions covering many experimentally defined candidate cis-regulatory elements. These databases are available on: [https://resources.aertslab.org/cistarget/](https://resources.aertslab.org/cistarget/).\n",
    "\n",
    "For this analysis we will use a precomputed database using [screen regions](https://screen.encodeproject.org/).\n",
    "\n",
    "Next to a precomputed motif database we also need a motif-to-tf annotation database. This is also available on [https://resources.aertslab.org/cistarget/](https://resources.aertslab.org/cistarget/)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load candidate enhancer regions identified in previous step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "region_bin_topics_otsu = pickle.load(open(os.path.join(work_dir, 'atac_ce3/candidate_enhancers/region_bin_topics_otsu_080825.pkl'), 'rb'))\n",
    "region_bin_topics_top3k = pickle.load(open(os.path.join(work_dir, 'atac_ce3/candidate_enhancers/region_bin_topics_top3k_080825.pkl'), 'rb'))\n",
    "markers_dict = pickle.load(open(os.path.join(work_dir, 'atac_ce3/candidate_enhancers/markers_dict_080825.pkl'), 'rb'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert to dictionary of pyranges objects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AMP':                           Log2FC  Adjusted_pval Contrast\n",
       " 4:54425323-54426024    10.157890  1.154359e-102      AMP\n",
       " 10:16431684-16432543    8.204540   4.693869e-42      AMP\n",
       " 1:59347241-59347944     7.559338  6.673294e-103      AMP\n",
       " 2:141156927-141157717   7.103826   3.138491e-16      AMP\n",
       " 1:115025593-115026480   7.054397  2.957856e-269      AMP\n",
       " ...                          ...            ...      ...\n",
       " 13:7300138-7300964      0.585054   6.049209e-93      AMP\n",
       " 5:36588888-36589751     0.585012  1.009051e-152      AMP\n",
       " 1:59255932-59256736     0.585010  7.702151e-127      AMP\n",
       " 12:11564714-11565594    0.584976   0.000000e+00      AMP\n",
       " 10:2163965-2164915      0.584975  9.770786e-154      AMP\n",
       " \n",
       " [25060 rows x 3 columns],\n",
       " 'DLL1-N':                         Log2FC  Adjusted_pval Contrast\n",
       " 3:78473862-78474551   7.195505   2.522113e-20   DLL1-N\n",
       " 12:8504639-8505471    5.696272   1.812261e-99   DLL1-N\n",
       " 1:70291766-70292239   5.554883   2.350091e-32   DLL1-N\n",
       " 4:61633732-61634604   5.125956   5.215853e-16   DLL1-N\n",
       " 3:25711029-25711515   4.942433   4.057503e-02   DLL1-N\n",
       " ...                        ...            ...      ...\n",
       " 4:33554282-33555108   0.585049   2.403776e-57   DLL1-N\n",
       " 3:35415883-35416791   0.585043  5.789239e-181   DLL1-N\n",
       " 10:17936413-17937260  0.585015  6.513842e-223   DLL1-N\n",
       " 4:5856350-5857369     0.585010   1.662284e-27   DLL1-N\n",
       " 8:12143631-12144449   0.584997  1.624653e-118   DLL1-N\n",
       " \n",
       " [20129 rows x 3 columns],\n",
       " 'DLX-N':                           Log2FC  Adjusted_pval Contrast\n",
       " 1:193389600-193390257  36.104914   9.327619e-03    DLX-N\n",
       " 4:51496197-51497085     6.235108   2.890483e-85    DLX-N\n",
       " Z:42978070-42979119     5.048436   4.755444e-99    DLX-N\n",
       " Z:2262059-2262380       4.997156  4.590016e-118    DLX-N\n",
       " 6:34368335-34369174     4.963488   2.485775e-14    DLX-N\n",
       " ...                          ...            ...      ...\n",
       " 5:13184754-13185636     0.585013  4.564039e-169    DLX-N\n",
       " 2:137725459-137726348   0.584999  6.067055e-170    DLX-N\n",
       " Z:7769298-7770073       0.584991   4.235601e-45    DLX-N\n",
       " 3:60426074-60426900     0.584988   6.347007e-65    DLX-N\n",
       " 1:146353555-146354401   0.584978   3.586144e-80    DLX-N\n",
       " \n",
       " [31024 rows x 3 columns],\n",
       " 'Die-FP':                          Log2FC  Adjusted_pval Contrast\n",
       " 3:80690060-80691050    4.093566   1.506055e-49   Die-FP\n",
       " 5:45433839-45434246    3.959312   5.807680e-70   Die-FP\n",
       " 1:66966828-66967462    3.955035   3.284640e-77   Die-FP\n",
       " Z:64668921-64669849    3.908520   1.807673e-49   Die-FP\n",
       " 2:144099344-144100197  3.890653   1.949043e-23   Die-FP\n",
       " ...                         ...            ...      ...\n",
       " 2:14633352-14634242    0.585217   9.959853e-33   Die-FP\n",
       " 13:4508477-4509061     0.585158   1.783884e-33   Die-FP\n",
       " 24:1244045-1244900     0.585153   4.859958e-47   Die-FP\n",
       " 2:142988133-142988922  0.585011   9.635780e-21   Die-FP\n",
       " 24:1827362-1828170     0.584980   9.573889e-36   Die-FP\n",
       " \n",
       " [22357 rows x 3 columns],\n",
       " 'Die-Mes-RP':                          Log2FC  Adjusted_pval    Contrast\n",
       " Z:32090912-32091543    7.084057   9.250518e-19  Die-Mes-RP\n",
       " 1:85190573-85191456    6.990155   1.494690e-15  Die-Mes-RP\n",
       " 3:62074846-62075637    6.288970   8.560538e-52  Die-Mes-RP\n",
       " 1:16399310-16400124    5.774526   8.367454e-09  Die-Mes-RP\n",
       " 2:97918987-97919895    5.564528  1.659312e-120  Die-Mes-RP\n",
       " ...                         ...            ...         ...\n",
       " 2:140563040-140563940  0.585043   1.938460e-57  Die-Mes-RP\n",
       " 2:47339676-47340512    0.585024  3.933955e-117  Die-Mes-RP\n",
       " 1:70177368-70178447    0.585023  4.127172e-105  Die-Mes-RP\n",
       " 2:23446664-23447424    0.585021  1.359012e-100  Die-Mes-RP\n",
       " 13:10321911-10322782   0.584967  1.595148e-101  Die-Mes-RP\n",
       " \n",
       " [32247 rows x 3 columns],\n",
       " 'GAP43-N':                           Log2FC  Adjusted_pval Contrast\n",
       " 1:61200242-61201111    35.434803   1.200217e-03  GAP43-N\n",
       " 6:18975943-18976742    10.080434   5.198759e-03  GAP43-N\n",
       " 3:35289892-35290371     8.995545   2.294040e-15  GAP43-N\n",
       " 23:2674617-2675061      8.100898   1.438042e-08  GAP43-N\n",
       " 3:65058292-65058970     8.042466   7.846281e-51  GAP43-N\n",
       " ...                          ...            ...      ...\n",
       " 3:109232889-109233515   0.585016   0.000000e+00  GAP43-N\n",
       " 7:23585872-23586728     0.584988  1.178785e-190  GAP43-N\n",
       " 21:2732950-2733780      0.584978   0.000000e+00  GAP43-N\n",
       " 25:11124-11854          0.584972   0.000000e+00  GAP43-N\n",
       " 20:9021704-9022471      0.584965  7.402013e-310  GAP43-N\n",
       " \n",
       " [38286 rows x 3 columns],\n",
       " 'Hem':                          Log2FC  Adjusted_pval Contrast\n",
       " 2:136704192-136704975  6.186963   1.576392e-06      Hem\n",
       " 2:15040554-15041023    6.042985   6.806805e-24      Hem\n",
       " 1:6638473-6639345      5.926549   9.827334e-17      Hem\n",
       " Z:12916145-12917047    5.920282   5.131299e-09      Hem\n",
       " 3:97492022-97492748    5.750515   5.450886e-17      Hem\n",
       " ...                         ...            ...      ...\n",
       " 1:141654092-141655024  0.585150   2.523028e-39      Hem\n",
       " 1:152959802-152960562  0.585150   3.266372e-62      Hem\n",
       " 5:49181889-49182634    0.585111   7.974091e-56      Hem\n",
       " 2:14233612-14234531    0.585110   6.290617e-21      Hem\n",
       " 11:10603430-10604352   0.585076   7.400931e-31      Hem\n",
       " \n",
       " [31201 rows x 3 columns],\n",
       " 'It-FP':                        Log2FC  Adjusted_pval Contrast\n",
       " 1:72725441-72726440  6.922429   5.670693e-07    It-FP\n",
       " 20:6089016-6089526   6.163739   3.240582e-10    It-FP\n",
       " 3:14966768-14967183  6.159778   1.357812e-55    It-FP\n",
       " 4:41708189-41709048  5.829028   5.439410e-29    It-FP\n",
       " 3:26104382-26104770  5.617702   1.486830e-31    It-FP\n",
       " ...                       ...            ...      ...\n",
       " 3:63726498-63727371  0.585159  5.607036e-124    It-FP\n",
       " 9:14653655-14654488  0.585124   6.272689e-62    It-FP\n",
       " 6:20026846-20027752  0.585119   2.680703e-62    It-FP\n",
       " 10:1828160-1829016   0.585096   8.998147e-54    It-FP\n",
       " 2:28044375-28045280  0.585007   2.564091e-17    It-FP\n",
       " \n",
       " [24824 rows x 3 columns],\n",
       " 'LHX3-N':                         Log2FC  Adjusted_pval Contrast\n",
       " 3:25711029-25711515   6.369133   7.881528e-03   LHX3-N\n",
       " 3:78473862-78474551   6.217130   2.957661e-09   LHX3-N\n",
       " 1:70291766-70292239   5.992315   1.648657e-22   LHX3-N\n",
       " 12:8504639-8505471    5.092802   3.828596e-21   LHX3-N\n",
       " 4:61633732-61634604   4.992865   3.791524e-05   LHX3-N\n",
       " ...                        ...            ...      ...\n",
       " 13:12337292-12338060  0.585048   1.071745e-15   LHX3-N\n",
       " 12:19709976-19710777  0.585021   1.265540e-11   LHX3-N\n",
       " Z:25821565-25822316   0.585014   5.403636e-10   LHX3-N\n",
       " 38:30333-31222        0.584988   4.319464e-21   LHX3-N\n",
       " 2:13762588-13763411   0.584983   5.962243e-13   LHX3-N\n",
       " \n",
       " [28769 rows x 3 columns],\n",
       " 'LHX9-N':                        Log2FC  Adjusted_pval Contrast\n",
       " Z:64003322-64004232  6.281346   3.161969e-26   LHX9-N\n",
       " 3:87055903-87056723  5.436620   4.077333e-45   LHX9-N\n",
       " 3:15997185-15998039  5.321846   1.247825e-30   LHX9-N\n",
       " Z:8590653-8591537    5.302278   7.217409e-99   LHX9-N\n",
       " 4:5662588-5663456    5.132921   3.663302e-04   LHX9-N\n",
       " ...                       ...            ...      ...\n",
       " 11:9035897-9036666   0.585152   2.459052e-07   LHX9-N\n",
       " 6:23831525-23832401  0.585107   7.003853e-81   LHX9-N\n",
       " 1:34055702-34056627  0.585036   9.045130e-49   LHX9-N\n",
       " 7:14797885-14798719  0.585018   8.185952e-53   LHX9-N\n",
       " 4:1293558-1294421    0.584967   3.931625e-99   LHX9-N\n",
       " \n",
       " [35657 rows x 3 columns],\n",
       " 'Mes-FP':                        Log2FC  Adjusted_pval Contrast\n",
       " Z:72370462-72371401  5.029590   6.330834e-58   Mes-FP\n",
       " 4:51981514-51982283  5.019346  3.486930e-185   Mes-FP\n",
       " Z:72724245-72725011  5.019006  3.657975e-199   Mes-FP\n",
       " 3:14684000-14684903  4.914527  3.225998e-247   Mes-FP\n",
       " 24:1414816-1415383   4.913746  3.290818e-139   Mes-FP\n",
       " ...                       ...            ...      ...\n",
       " 4:43244046-43244765  0.585109  4.781222e-229   Mes-FP\n",
       " 14:7134994-7135864   0.585100   3.602326e-72   Mes-FP\n",
       " 3:13928205-13929097  0.585055   1.103532e-23   Mes-FP\n",
       " 4:38941501-38942368  0.585031  2.436875e-155   Mes-FP\n",
       " 2:23782910-23783695  0.584964  1.022551e-142   Mes-FP\n",
       " \n",
       " [20586 rows x 3 columns],\n",
       " 'OV':                          Log2FC  Adjusted_pval Contrast\n",
       " 4:81495467-81496161    6.715258   5.055478e-73       OV\n",
       " 1:156998636-156999467  6.135406   2.922749e-04       OV\n",
       " 11:16921870-16922515   6.117445   8.060159e-10       OV\n",
       " 2:101670730-101671506  5.533737   0.000000e+00       OV\n",
       " 4:64001494-64002390    5.400077   0.000000e+00       OV\n",
       " ...                         ...            ...      ...\n",
       " 3:98969354-98970207    0.585050  2.626663e-143       OV\n",
       " 3:93061513-93062432    0.585050  2.200931e-186       OV\n",
       " 10:12508454-12509254   0.585040  5.929840e-110       OV\n",
       " 30:415353-416183       0.584987  8.903307e-203       OV\n",
       " 28:1319131-1319842     0.584977  1.469038e-261       OV\n",
       " \n",
       " [15736 rows x 3 columns],\n",
       " 'P1':                        Log2FC  Adjusted_pval Contrast\n",
       " 4:63947616-63948499  7.420130   2.294543e-03       P1\n",
       " 3:2834929-2835424    5.417242  9.054533e-140       P1\n",
       " 4:32142657-32143560  5.411681  5.468727e-116       P1\n",
       " Z:12305246-12306154  5.161212   5.842106e-15       P1\n",
       " 1:62491655-62492490  4.931092   3.654074e-11       P1\n",
       " ...                       ...            ...      ...\n",
       " 7:18132123-18132998  0.585171  4.503479e-166       P1\n",
       " 5:47400470-47401310  0.585147  5.290918e-100       P1\n",
       " 3:48270084-48271030  0.585111  3.016273e-307       P1\n",
       " 5:33752319-33753207  0.585100  5.104384e-155       P1\n",
       " Z:61871915-61872691  0.584994   8.896334e-16       P1\n",
       " \n",
       " [12319 rows x 3 columns],\n",
       " 'P3':                          Log2FC  Adjusted_pval Contrast\n",
       " 11:4472478-4473404     3.983630   4.087844e-03       P3\n",
       " 6:19944775-19945601    3.567543   3.375482e-11       P3\n",
       " 10:11300454-11301327   3.486061   2.314892e-23       P3\n",
       " 3:5169307-5170121      3.266802   9.021329e-36       P3\n",
       " 2:119935827-119936682  3.202097  6.233010e-166       P3\n",
       " ...                         ...            ...      ...\n",
       " 9:7496388-7496904      0.585172  2.077651e-103       P3\n",
       " 5:39949067-39949965    0.585169  3.678092e-125       P3\n",
       " 2:140941316-140942244  0.585151  1.942234e-250       P3\n",
       " 5:12404809-12405590    0.585101   1.629797e-10       P3\n",
       " 13:17372469-17373214   0.585081  5.240317e-120       P3\n",
       " \n",
       " [15900 rows x 3 columns],\n",
       " 'PG':                          Log2FC  Adjusted_pval Contrast\n",
       " 2:99306064-99306607    7.197537   1.926055e-06       PG\n",
       " Z:32090912-32091543    6.544759   2.133702e-04       PG\n",
       " 1:106413845-106414711  6.138843   8.907599e-09       PG\n",
       " 1:148347010-148347928  5.526301   2.753132e-06       PG\n",
       " 1:6638473-6639345      5.391915   2.274006e-02       PG\n",
       " ...                         ...            ...      ...\n",
       " 1:80799068-80799909    0.585004   2.203352e-03       PG\n",
       " 5:18839310-18840201    0.585000   6.699543e-05       PG\n",
       " 5:8490113-8490929      0.584998   8.487767e-05       PG\n",
       " 4:60178555-60179397    0.584973   1.436012e-02       PG\n",
       " 12:7095566-7096458     0.584968   9.784337e-10       PG\n",
       " \n",
       " [34819 rows x 3 columns],\n",
       " 'POA':                         Log2FC  Adjusted_pval Contrast\n",
       " 5:11916376-11916833  10.497568   1.398975e-02      POA\n",
       " 19:7802361-7803207    9.795368   2.180520e-52      POA\n",
       " 8:25339613-25340518   6.759744  8.059408e-114      POA\n",
       " 1:45756621-45757384   6.632964  2.967536e-262      POA\n",
       " 4:49660705-49661437   5.481371   0.000000e+00      POA\n",
       " ...                        ...            ...      ...\n",
       " 9:23185426-23186282   0.585333  5.360207e-126      POA\n",
       " 4:4158326-4159182     0.585229   6.914257e-86      POA\n",
       " 3:54520448-54521288   0.585015   1.315109e-44      POA\n",
       " 19:6048134-6049101    0.585001  1.000641e-178      POA\n",
       " Z:23284580-23285466   0.584965  5.740657e-101      POA\n",
       " \n",
       " [16326 rows x 3 columns],\n",
       " 'RL':                         Log2FC  Adjusted_pval Contrast\n",
       " 6:19628513-19629144   7.372979   1.285314e-58       RL\n",
       " 4:78169677-78170115   6.579192   3.742574e-03       RL\n",
       " 1:40050137-40050977   6.106393   1.484867e-09       RL\n",
       " 3:98791039-98791947   5.848103   1.622489e-02       RL\n",
       " 4:65557117-65557993   5.727443  2.022132e-155       RL\n",
       " ...                        ...            ...      ...\n",
       " 3:44970740-44971665   0.585246  1.865787e-143       RL\n",
       " 4:5987614-5988483     0.585180  3.488194e-127       RL\n",
       " 10:7447330-7448216    0.585157   0.000000e+00       RL\n",
       " 10:17255481-17256343  0.585048  6.416132e-121       RL\n",
       " 11:10957924-10958773  0.585032  1.049874e-169       RL\n",
       " \n",
       " [19403 rows x 3 columns],\n",
       " 'Rho-RP':                           Log2FC  Adjusted_pval Contrast\n",
       " 1:70552117-70552868    10.343377   2.217434e-13   Rho-RP\n",
       " 2:56030841-56031650     9.853466   8.644534e-14   Rho-RP\n",
       " 1:174352553-174352943   8.762237   6.084408e-06   Rho-RP\n",
       " Z:41793078-41793844     8.552138   1.626645e-23   Rho-RP\n",
       " 6:19821585-19822342     8.157400   7.203485e-54   Rho-RP\n",
       " ...                          ...            ...      ...\n",
       " 3:82044586-82045400     0.585012   2.468120e-41   Rho-RP\n",
       " 18:6549561-6550392      0.585004   2.298458e-30   Rho-RP\n",
       " 9:15867557-15868305     0.584981   2.812474e-12   Rho-RP\n",
       " 2:61407847-61408660     0.584979   1.864198e-10   Rho-RP\n",
       " 8:13322959-13323899     0.584973   1.169246e-17   Rho-RP\n",
       " \n",
       " [34368 rows x 3 columns],\n",
       " 'Rho-dFP':                          Log2FC  Adjusted_pval Contrast\n",
       " 12:4667432-4668304     5.218393   2.572106e-03  Rho-dFP\n",
       " 4:60263194-60264111    4.040601   1.294904e-26  Rho-dFP\n",
       " 8:4902092-4903005      3.782567   6.210830e-37  Rho-dFP\n",
       " 2:19792128-19792992    3.772611   2.536188e-41  Rho-dFP\n",
       " 3:49959324-49959897    3.678993   3.784555e-61  Rho-dFP\n",
       " ...                         ...            ...      ...\n",
       " 7:18095870-18096632    0.585051   4.849073e-25  Rho-dFP\n",
       " 4:52644026-52644873    0.585051   4.900690e-37  Rho-dFP\n",
       " 7:9586249-9587218      0.585050  1.391136e-135  Rho-dFP\n",
       " 1:110649475-110650385  0.585037   1.052695e-91  Rho-dFP\n",
       " 3:64648417-64649159    0.584993   2.739360e-37  Rho-dFP\n",
       " \n",
       " [19453 rows x 3 columns],\n",
       " 'Rho-vFP':                          Log2FC  Adjusted_pval Contrast\n",
       " 2:126331505-126331902  7.402455   2.620175e-57  Rho-vFP\n",
       " 20:6089016-6089526     6.707554   3.223591e-15  Rho-vFP\n",
       " 1:79069652-79070586    6.424891   5.525127e-94  Rho-vFP\n",
       " 2:116834501-116834854  6.418407   5.649532e-94  Rho-vFP\n",
       " 2:21602057-21603041    6.338651   2.794100e-76  Rho-vFP\n",
       " ...                         ...            ...      ...\n",
       " 9:12796165-12796849    0.585161   4.100314e-27  Rho-vFP\n",
       " 1:9036838-9037742      0.585160   4.527403e-42  Rho-vFP\n",
       " 5:13498776-13499664    0.585094   1.440018e-47  Rho-vFP\n",
       " 8:14134736-14135604    0.585009  1.899242e-147  Rho-vFP\n",
       " 17:9737402-9738138     0.585004   2.889195e-83  Rho-vFP\n",
       " \n",
       " [27921 rows x 3 columns],\n",
       " 'TAL1-N':                        Log2FC  Adjusted_pval Contrast\n",
       " Z:66235618-66236489  9.597481   2.682815e-10   TAL1-N\n",
       " 10:7117642-7118129   7.212543   1.161622e-22   TAL1-N\n",
       " 6:34368335-34369174  6.707664   1.098139e-52   TAL1-N\n",
       " 13:9361326-9362138   6.138332   1.081578e-30   TAL1-N\n",
       " 8:25153730-25154484  5.879960   3.277573e-18   TAL1-N\n",
       " ...                       ...            ...      ...\n",
       " 6:2651152-2652019    0.585015   8.862777e-26   TAL1-N\n",
       " 26:3369639-3370571   0.585011   1.350043e-46   TAL1-N\n",
       " 5:30437894-30438879  0.584990   3.507959e-31   TAL1-N\n",
       " Z:35824522-35825429  0.584989   6.470054e-38   TAL1-N\n",
       " 1:54010562-54011362  0.584983   9.934464e-47   TAL1-N\n",
       " \n",
       " [31709 rows x 3 columns],\n",
       " 'TAL2-N':                          Log2FC  Adjusted_pval Contrast\n",
       " 6:24482182-24482548    7.147306   4.612192e-02   TAL2-N\n",
       " 6:22714850-22715644    5.864589   3.960485e-09   TAL2-N\n",
       " 4:12599900-12600803    5.667040   9.982713e-20   TAL2-N\n",
       " 1:29229263-29230062    5.454238   1.763645e-26   TAL2-N\n",
       " 3:2834929-2835424      5.232194   2.139179e-30   TAL2-N\n",
       " ...                         ...            ...      ...\n",
       " 3:92619100-92619935    0.585052   1.871809e-08   TAL2-N\n",
       " 19:884274-885186       0.585047   1.095427e-10   TAL2-N\n",
       " 1:146445055-146445995  0.584997   5.257525e-12   TAL2-N\n",
       " Z:31349137-31350019    0.584978   2.513288e-08   TAL2-N\n",
       " 1:21935383-21936303    0.584964   3.298463e-12   TAL2-N\n",
       " \n",
       " [18689 rows x 3 columns],\n",
       " 'TBX20-N':                         Log2FC  Adjusted_pval Contrast\n",
       " 4:61633732-61634604   3.394190   1.711696e-02  TBX20-N\n",
       " 11:18159060-18159384  3.344608   2.325916e-55  TBX20-N\n",
       " 12:5940352-5941050    3.217312   2.786905e-09  TBX20-N\n",
       " 9:14021076-14021483   3.213865   3.444568e-12  TBX20-N\n",
       " Z:50994596-50995537   3.189269   1.839295e-58  TBX20-N\n",
       " ...                        ...            ...      ...\n",
       " 13:10134068-10134934  0.585131   1.446134e-49  TBX20-N\n",
       " 1:96107163-96107930   0.585078   3.483715e-46  TBX20-N\n",
       " 13:15293256-15294163  0.585010   5.566421e-18  TBX20-N\n",
       " 26:5005205-5005912    0.584974   6.011634e-97  TBX20-N\n",
       " 8:429628-430548       0.584971   1.645121e-41  TBX20-N\n",
       " \n",
       " [37145 rows x 3 columns],\n",
       " 'THT-PHT':                          Log2FC  Adjusted_pval Contrast\n",
       " 2:68073663-68074506    5.799488  1.187963e-226  THT-PHT\n",
       " 10:14362742-14363613   5.708904  9.966861e-185  THT-PHT\n",
       " Z:42978070-42979119    5.672753  3.550437e-213  THT-PHT\n",
       " 5:50462719-50463425    5.629924  8.037831e-209  THT-PHT\n",
       " 8:16746522-16747274    5.560596  9.839014e-182  THT-PHT\n",
       " ...                         ...            ...      ...\n",
       " 2:24178929-24179639    0.585017  8.643430e-128  THT-PHT\n",
       " 4:29815158-29816021    0.585015  2.146603e-139  THT-PHT\n",
       " 3:105800006-105800868  0.585007   3.179721e-80  THT-PHT\n",
       " 2:107581768-107582535  0.585001  1.109894e-292  THT-PHT\n",
       " 5:18865809-18866444    0.584984  1.655057e-171  THT-PHT\n",
       " \n",
       " [21682 rows x 3 columns],\n",
       " \"a'It\":                           Log2FC  Adjusted_pval Contrast\n",
       " Z:67646217-67647238    10.075738   1.550030e-47     a'It\n",
       " Z:56585468-56586328     8.948159   1.080366e-13     a'It\n",
       " 2:55684208-55685089     7.865977   6.795952e-55     a'It\n",
       " 4:5772439-5773042       7.431302   1.153564e-52     a'It\n",
       " 13:13816046-13816942    6.603397  4.565810e-130     a'It\n",
       " ...                          ...            ...      ...\n",
       " 4:45847252-45848116     0.585057   1.835466e-77     a'It\n",
       " 1:189568548-189569419   0.585036  5.732729e-128     a'It\n",
       " 5:46939603-46940422     0.585032   2.508396e-80     a'It\n",
       " 2:104496662-104497551   0.584987  4.997681e-107     a'It\n",
       " 8:24875928-24876762     0.584974  9.710053e-195     a'It\n",
       " \n",
       " [20983 rows x 3 columns],\n",
       " \"a'lMes\":                          Log2FC  Adjusted_pval Contrast\n",
       " Z:61805135-61805849    7.746385   3.155227e-23   a'lMes\n",
       " 2:27786171-27787057    7.260381   4.040930e-21   a'lMes\n",
       " 1:74214714-74215463    6.263116   3.283057e-74   a'lMes\n",
       " 2:118189809-118190730  6.208719   1.353676e-78   a'lMes\n",
       " 4:30936903-30937286    5.352762   4.113642e-48   a'lMes\n",
       " ...                         ...            ...      ...\n",
       " 4:21007488-21008361    0.585143   9.095303e-87   a'lMes\n",
       " 8:17765844-17766645    0.585134   4.167683e-26   a'lMes\n",
       " 2:73865524-73866268    0.585112   1.191359e-25   a'lMes\n",
       " 1:72587429-72588360    0.585105   2.528081e-83   a'lMes\n",
       " 8:2541790-2542693      0.585018   2.027792e-84   a'lMes\n",
       " \n",
       " [16572 rows x 3 columns],\n",
       " 'aMes':                          Log2FC  Adjusted_pval Contrast\n",
       " 7:24164251-24164935    5.574339   0.000000e+00     aMes\n",
       " 5:58081930-58082620    4.781000   0.000000e+00     aMes\n",
       " 1:133680376-133681120  4.339319   0.000000e+00     aMes\n",
       " 1:62613554-62614530    4.104256   5.214439e-90     aMes\n",
       " 7:11502994-11503310    3.960327   0.000000e+00     aMes\n",
       " ...                         ...            ...      ...\n",
       " 1:134709599-134710512  0.585224   1.351052e-95     aMes\n",
       " 1:95050115-95051029    0.585181  4.031845e-269     aMes\n",
       " Z:20708325-20708850    0.585152   3.298786e-70     aMes\n",
       " 4:18178167-18179050    0.585119   0.000000e+00     aMes\n",
       " 9:16711767-16712672    0.585091   0.000000e+00     aMes\n",
       " \n",
       " [16833 rows x 3 columns],\n",
       " 'aP2':                        Log2FC  Adjusted_pval Contrast\n",
       " Z:24479306-24479744  7.241807   2.070073e-02      aP2\n",
       " Z:10180099-10180955  7.083545   2.723021e-13      aP2\n",
       " 1:44727513-44728373  6.570871   6.373349e-05      aP2\n",
       " 3:99382030-99382767  5.747610   1.500454e-83      aP2\n",
       " 6:19944775-19945601  5.245943   2.546276e-42      aP2\n",
       " ...                       ...            ...      ...\n",
       " 13:9524858-9525773   0.585161  7.608512e-130      aP2\n",
       " 3:93553008-93553519  0.585050  3.764132e-131      aP2\n",
       " 1:23188508-23189181  0.585022  9.411122e-241      aP2\n",
       " 5:20612205-20612902  0.585002  2.240702e-128      aP2\n",
       " 6:24276009-24276881  0.584992  3.933955e-121      aP2\n",
       " \n",
       " [24399 rows x 3 columns],\n",
       " 'aR1-2':                         Log2FC  Adjusted_pval Contrast\n",
       " 4:3130621-3131496     7.506209   1.292544e-12    aR1-2\n",
       " 2:47631336-47632256   7.024227   9.000797e-94    aR1-2\n",
       " 5:50614292-50615004   6.987200   1.634789e-05    aR1-2\n",
       " 8:21303537-21304130   5.985238   6.935107e-11    aR1-2\n",
       " 5:18966466-18966886   5.931969   7.073427e-79    aR1-2\n",
       " ...                        ...            ...      ...\n",
       " 2:26147258-26147985   0.585123  1.640841e-122    aR1-2\n",
       " 15:3667913-3668698    0.585015  5.036215e-103    aR1-2\n",
       " 6:1949998-1950734     0.585011   4.601827e-31    aR1-2\n",
       " 11:10941262-10942038  0.584987   4.330673e-58    aR1-2\n",
       " 2:48818651-48819529   0.584970  1.306824e-133    aR1-2\n",
       " \n",
       " [13556 rows x 3 columns],\n",
       " 'aR3-8':                          Log2FC  Adjusted_pval Contrast\n",
       " 1:38835340-38836071    9.034676   2.173009e-04    aR3-8\n",
       " 27:2042171-2042978     7.636513   7.375760e-21    aR3-8\n",
       " 3:28552804-28553485    7.292042   2.010297e-28    aR3-8\n",
       " 1:180776362-180776772  5.660051   2.913475e-47    aR3-8\n",
       " 3:72334173-72334890    5.601038   1.960647e-91    aR3-8\n",
       " ...                         ...            ...      ...\n",
       " 7:5021963-5022866      0.585061   1.011535e-85    aR3-8\n",
       " 1:104910433-104911290  0.585061   2.212992e-75    aR3-8\n",
       " 1:12760671-12761343    0.585021  1.492631e-177    aR3-8\n",
       " Z:34479778-34480516    0.585019  2.115204e-198    aR3-8\n",
       " 2:86396950-86397825    0.584988   2.503558e-84    aR3-8\n",
       " \n",
       " [17548 rows x 3 columns],\n",
       " 'bMes':                          Log2FC  Adjusted_pval Contrast\n",
       " 13:15077715-15078569   4.658874   2.109062e-24     bMes\n",
       " 2:119507663-119508555  4.428187   3.801726e-48     bMes\n",
       " 3:72191391-72191846    3.785888  2.182450e-189     bMes\n",
       " Z:59438205-59439100    3.695440  1.672712e-144     bMes\n",
       " 3:28160807-28161614    3.693564   2.595946e-86     bMes\n",
       " ...                         ...            ...      ...\n",
       " 27:5132023-5132841     0.585349   5.347604e-74     bMes\n",
       " 7:2693526-2694405      0.585276   3.170976e-43     bMes\n",
       " 3:77269295-77270183    0.585185  1.607549e-132     bMes\n",
       " 2:17835346-17836237    0.585182  1.349021e-104     bMes\n",
       " 24:2239003-2239913     0.585061   2.068648e-98     bMes\n",
       " \n",
       " [13225 rows x 3 columns],\n",
       " 'bP2':                          Log2FC  Adjusted_pval Contrast\n",
       " 1:30802071-30802745    4.916689   1.191556e-05      bP2\n",
       " 2:119507663-119508555  3.950535   1.635765e-31      bP2\n",
       " 3:99382030-99382767    3.809424   4.107618e-18      bP2\n",
       " 1:89830776-89831321    3.769378   5.505288e-40      bP2\n",
       " Z:16483227-16484133    3.744789   1.276671e-06      bP2\n",
       " ...                         ...            ...      ...\n",
       " 3:17197666-17198569    0.585125  5.650177e-186      bP2\n",
       " 5:55403468-55404277    0.585110   5.081571e-68      bP2\n",
       " 19:9331437-9332182     0.585066  1.384784e-112      bP2\n",
       " 3:26060657-26061547    0.584996  3.662959e-108      bP2\n",
       " 1:34650539-34651250    0.584966   8.521228e-49      bP2\n",
       " \n",
       " [13455 rows x 3 columns],\n",
       " 'bR1-8':                          Log2FC  Adjusted_pval Contrast\n",
       " 12:9002285-9002787     6.713201   4.082292e-21    bR1-8\n",
       " 2:118091396-118092207  6.248404   9.412556e-41    bR1-8\n",
       " 2:17369222-17370051    5.224643  2.601198e-195    bR1-8\n",
       " 10:11420587-11421389   5.129770   4.361799e-27    bR1-8\n",
       " 4:52621673-52622163    4.449847  1.453116e-170    bR1-8\n",
       " ...                         ...            ...      ...\n",
       " Z:49166635-49167435    0.585131   1.699591e-90    bR1-8\n",
       " 1:85731340-85731773    0.585131   9.196236e-59    bR1-8\n",
       " 19:5083180-5084090     0.585108  7.145438e-140    bR1-8\n",
       " 8:13491120-13492027    0.585053  1.759785e-248    bR1-8\n",
       " Z:84799752-84800639    0.584980   7.715037e-26    bR1-8\n",
       " \n",
       " [15306 rows x 3 columns],\n",
       " 'dTel':                          Log2FC  Adjusted_pval Contrast\n",
       " 2:75681541-75682063    8.084336   5.478576e-04     dTel\n",
       " 3:110014358-110015198  7.050327  8.942625e-152     dTel\n",
       " 1:110892063-110892807  6.419440   0.000000e+00     dTel\n",
       " 8:2273102-2273857      6.252205  4.218682e-247     dTel\n",
       " 6:15486487-15487108    6.026165  4.167179e-157     dTel\n",
       " ...                         ...            ...      ...\n",
       " 1:171790956-171791768  0.585274  7.693384e-190     dTel\n",
       " Z:47332580-47333476    0.585179  3.147823e-219     dTel\n",
       " 25:244257-244985       0.585055   0.000000e+00     dTel\n",
       " 2:104455568-104456461  0.585041  8.666911e-201     dTel\n",
       " 9:3929295-3930132      0.584989  8.442508e-299     dTel\n",
       " \n",
       " [17667 rows x 3 columns],\n",
       " 'lR1-8':                          Log2FC  Adjusted_pval Contrast\n",
       " 12:19982519-19983372   3.832892   3.961383e-25    lR1-8\n",
       " 2:124586880-124587922  3.486482  2.052637e-179    lR1-8\n",
       " 3:53366657-53367546    3.078129   1.312425e-02    lR1-8\n",
       " 2:17369222-17370051    2.999810   8.702007e-32    lR1-8\n",
       " Z:22586695-22587451    2.888919  8.527898e-152    lR1-8\n",
       " ...                         ...            ...      ...\n",
       " 6:26649475-26650360    0.585084   0.000000e+00    lR1-8\n",
       " 4:78838166-78839014    0.585048   3.444322e-85    lR1-8\n",
       " 1:148008893-148009701  0.584978  1.176444e-294    lR1-8\n",
       " 7:5036899-5037818      0.584975  5.654687e-233    lR1-8\n",
       " 10:3155327-3156245     0.584968   0.000000e+00    lR1-8\n",
       " \n",
       " [13314 rows x 3 columns],\n",
       " 'pIt':                          Log2FC  Adjusted_pval Contrast\n",
       " 1:193448319-193449177  8.722301   7.303042e-13      pIt\n",
       " 6:19809673-19810206    8.412446   4.133296e-03      pIt\n",
       " 3:97495603-97496460    6.498820   7.527830e-57      pIt\n",
       " 2:139025503-139026408  6.429519  1.167620e-121      pIt\n",
       " 1:152679930-152680473  6.389020   1.319253e-63      pIt\n",
       " ...                         ...            ...      ...\n",
       " 26:2461071-2461919     0.585205  7.603845e-253      pIt\n",
       " 19:5034843-5035692     0.585198  3.191624e-104      pIt\n",
       " 5:24447290-24447976    0.585123  3.661519e-171      pIt\n",
       " 2:147305462-147306342  0.585068  3.568337e-115      pIt\n",
       " 5:7809206-7810022      0.585061  1.074366e-130      pIt\n",
       " \n",
       " [19056 rows x 3 columns],\n",
       " 'plMes':                        Log2FC  Adjusted_pval Contrast\n",
       " 2:40207076-40207379  4.983363   5.056399e-65    plMes\n",
       " 6:34825018-34825809  4.455505   4.393841e-59    plMes\n",
       " 15:8370383-8371061   4.150376  1.146404e-300    plMes\n",
       " 2:24869898-24870048  3.970340   3.083710e-96    plMes\n",
       " Z:8832610-8833150    3.783999  2.118494e-176    plMes\n",
       " ...                       ...            ...      ...\n",
       " 19:7820591-7821092   0.585187   0.000000e+00    plMes\n",
       " 2:83349082-83350202  0.585143  1.882560e-227    plMes\n",
       " 4:6091370-6092247    0.585134   4.584624e-66    plMes\n",
       " 3:99084612-99085506  0.585128  4.637795e-223    plMes\n",
       " 11:1298538-1299455   0.585070   0.000000e+00    plMes\n",
       " \n",
       " [12403 rows x 3 columns],\n",
       " 'vTel':                          Log2FC  Adjusted_pval Contrast\n",
       " 1:72167501-72168270    5.930473   0.000000e+00     vTel\n",
       " 1:134184961-134185695  5.485005   0.000000e+00     vTel\n",
       " 2:100611093-100611974  5.453209  1.054236e-227     vTel\n",
       " 1:107785770-107786578  5.097992   0.000000e+00     vTel\n",
       " 6:22714850-22715644    4.527960   2.724415e-10     vTel\n",
       " ...                         ...            ...      ...\n",
       " 4:76168959-76169947    0.585112  2.687172e-229     vTel\n",
       " 15:10256079-10256983   0.585092  1.204791e-194     vTel\n",
       " 3:58759654-58760490    0.585063   7.019968e-43     vTel\n",
       " 5:30716658-30717406    0.585051   0.000000e+00     vTel\n",
       " 10:16714904-16715749   0.584984  1.227863e-126     vTel\n",
       " \n",
       " [16060 rows x 3 columns]}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "markers_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1:132793951-132794867</th>\n",
       "      <td>1</td>\n",
       "      <td>132793951</td>\n",
       "      <td>132794867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27:2422137-2423009</th>\n",
       "      <td>27</td>\n",
       "      <td>2422137</td>\n",
       "      <td>2423009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10:2534463-2535390</th>\n",
       "      <td>10</td>\n",
       "      <td>2534463</td>\n",
       "      <td>2535390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4:81624789-81625649</th>\n",
       "      <td>4</td>\n",
       "      <td>81624789</td>\n",
       "      <td>81625649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:14345689-14346556</th>\n",
       "      <td>1</td>\n",
       "      <td>14345689</td>\n",
       "      <td>14346556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4:58194043-58194922</th>\n",
       "      <td>4</td>\n",
       "      <td>58194043</td>\n",
       "      <td>58194922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27:3881257-3881833</th>\n",
       "      <td>27</td>\n",
       "      <td>3881257</td>\n",
       "      <td>3881833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:187876076-187876994</th>\n",
       "      <td>1</td>\n",
       "      <td>187876076</td>\n",
       "      <td>187876994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:29781883-29782817</th>\n",
       "      <td>1</td>\n",
       "      <td>29781883</td>\n",
       "      <td>29782817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2:38875501-38876408</th>\n",
       "      <td>2</td>\n",
       "      <td>38875501</td>\n",
       "      <td>38876408</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6024 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Chromosome      Start        End\n",
       "1:132793951-132794867          1  132793951  132794867\n",
       "27:2422137-2423009            27    2422137    2423009\n",
       "10:2534463-2535390            10    2534463    2535390\n",
       "4:81624789-81625649            4   81624789   81625649\n",
       "1:14345689-14346556            1   14345689   14346556\n",
       "...                          ...        ...        ...\n",
       "4:58194043-58194922            4   58194043   58194922\n",
       "27:3881257-3881833            27    3881257    3881833\n",
       "1:187876076-187876994          1  187876076  187876994\n",
       "1:29781883-29782817            1   29781883   29782817\n",
       "2:38875501-38876408            2   38875501   38876408\n",
       "\n",
       "[6024 rows x 3 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions = region_bin_topics_otsu[\"Topic1\"].index\n",
    "region_names_to_coordinates(regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "from pycistarget.utils import region_names_to_coordinates\n",
    "region_sets = {}\n",
    "region_sets['topics_otsu'] = {}\n",
    "region_sets['topics_top_3'] = {}\n",
    "region_sets['DARs'] = {}\n",
    "for topic in region_bin_topics_otsu.keys():\n",
    "    regions = region_bin_topics_otsu[topic].index #only keep regions on known chromosomes\n",
    "    region_sets['topics_otsu'][topic] = pr.PyRanges(region_names_to_coordinates(regions))\n",
    "for topic in region_bin_topics_top3k.keys():\n",
    "    regions = region_bin_topics_top3k[topic].index #only keep regions on known chromosomes\n",
    "    region_sets['topics_top_3'][topic] = pr.PyRanges(region_names_to_coordinates(regions))\n",
    "for DAR in markers_dict.keys():\n",
    "    regions = markers_dict[DAR].index #only keep regions on known chromosomes\n",
    "    region_sets['DARs'][DAR] = pr.PyRanges(region_names_to_coordinates(regions))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "topics_otsu: dict_keys(['Topic1', 'Topic2', 'Topic3', 'Topic4', 'Topic5', 'Topic6', 'Topic7', 'Topic8', 'Topic9', 'Topic10', 'Topic11', 'Topic12', 'Topic13', 'Topic14', 'Topic15', 'Topic16', 'Topic17', 'Topic18', 'Topic19', 'Topic20', 'Topic21', 'Topic22', 'Topic23', 'Topic24', 'Topic25', 'Topic26', 'Topic27', 'Topic28', 'Topic29', 'Topic30', 'Topic31', 'Topic32', 'Topic33', 'Topic34', 'Topic35', 'Topic36', 'Topic37', 'Topic38', 'Topic39', 'Topic40', 'Topic41', 'Topic42', 'Topic43', 'Topic44', 'Topic45', 'Topic46', 'Topic47', 'Topic48', 'Topic49', 'Topic50'])\n",
      "topics_top_3: dict_keys(['Topic1', 'Topic2', 'Topic3', 'Topic4', 'Topic5', 'Topic6', 'Topic7', 'Topic8', 'Topic9', 'Topic10', 'Topic11', 'Topic12', 'Topic13', 'Topic14', 'Topic15', 'Topic16', 'Topic17', 'Topic18', 'Topic19', 'Topic20', 'Topic21', 'Topic22', 'Topic23', 'Topic24', 'Topic25', 'Topic26', 'Topic27', 'Topic28', 'Topic29', 'Topic30', 'Topic31', 'Topic32', 'Topic33', 'Topic34', 'Topic35', 'Topic36', 'Topic37', 'Topic38', 'Topic39', 'Topic40', 'Topic41', 'Topic42', 'Topic43', 'Topic44', 'Topic45', 'Topic46', 'Topic47', 'Topic48', 'Topic49', 'Topic50'])\n",
      "DARs: dict_keys(['AMP', 'DLL1-N', 'DLX-N', 'Die-FP', 'Die-Mes-RP', 'GAP43-N', 'Hem', 'It-FP', 'LHX3-N', 'LHX9-N', 'Mes-FP', 'OV', 'P1', 'P3', 'PG', 'POA', 'RL', 'Rho-RP', 'Rho-dFP', 'Rho-vFP', 'TAL1-N', 'TAL2-N', 'TBX20-N', 'THT-PHT', \"a'It\", \"a'lMes\", 'aMes', 'aP2', 'aR1-2', 'aR3-8', 'bMes', 'bP2', 'bR1-8', 'dTel', 'lR1-8', 'pIt', 'plMes', 'vTel'])\n"
     ]
    }
   ],
   "source": [
    "for key in region_sets.keys():\n",
    "    print(f'{key}: {region_sets[key].keys()}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define rankings, score and motif annotation database.\n",
    "\n",
    "The ranking database is used for running the cistarget analysis and the scores database is used for running the DEM analysis. For more information see [the pycistarget read the docs page](https://pycistarget.readthedocs.io/en/latest/)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "db_fpath = \"/work/rsenovilla/scenic/cistarget\"\n",
    "motif_annot_fpath = \"/work/rsenovilla/scenic/cistarget\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "rankings_db = os.path.join(db_fpath, 'chickBrain_combined.regions_vs_motifs.rankings.feather')\n",
    "rankings_db_mod = os.path.join(db_fpath, 'rsg_chickBrain_combined.regions_vs_motifs.rankings.feather')\n",
    "scores_db =  os.path.join(db_fpath, 'chickBrain_combined.regions_vs_motifs.scores.feather')\n",
    "scores_db_mod =  os.path.join(db_fpath, 'rsg_chickBrain_combined.regions_vs_motifs.scores.feather')\n",
    "motif_annotation = os.path.join(motif_annot_fpath, 'motifs-v10nr_clust-nr.chicken-m0.001-o0.0.tbl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we will run pycistarget using the `run_pycistarget` wrapper function.\n",
    "\n",
    "This function will run cistarget based and DEM based motif enrichment analysis with or without promoter regions.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(os.path.join(work_dir, 'atac_ce3/motifs_080825')):\n",
    "    os.makedirs(os.path.join(work_dir, 'atac_ce3/motifs_080825'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "topics_otsu: 50 regions\n",
      "topics_top_3: 50 regions\n",
      "DARs: 38 regions\n"
     ]
    }
   ],
   "source": [
    "for name, regions in region_sets.items():\n",
    "    print(f\"{name}: {len(regions)} regions\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "rkn = pd.read_feather(rankings_db)\n",
    "rkn = rkn.rename(columns=lambda x: x.replace('chr', ''))\n",
    "rkn.to_feather(rankings_db_mod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scores = pd.read_feather(scores_db)\n",
    "scores = scores.rename(columns=lambda x: x.replace('chr', ''))\n",
    "scores.to_feather(scores_db_mod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scenicplus.wrappers.run_pycistarget import run_pycistarget\n",
    "run_pycistarget(\n",
    "    region_sets = region_sets,\n",
    "    species = 'gallus_gallus',\n",
    "    save_path = os.path.join(work_dir, 'atac_ce3/motifs_080825'),\n",
    "    ctx_db_path = rankings_db_mod,\n",
    "    dem_db_path = scores_db_mod,\n",
    "    path_to_motif_annotations = motif_annotation,\n",
    "    run_without_promoters = True,\n",
    "    n_cpu = 20,\n",
    "    _temp_dir = os.path.join(tmp_dir, 'ray_spill'),\n",
    "    annotation_version = 'v10nr_clust',\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pickle.dump(cistopic_obj,\n",
    "          open(os.path.join(work_dir, 'atac_ce3/cistopic_obj_080825.pkl'), 'wb'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's explore some of the results. Below we show the motifs found for topic 8 (specific to B-cells) using DEM."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dill\n",
    "menr = dill.load(open(os.path.join(work_dir, 'atac_ce3/motifs_080825/menr.pkl'), 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'menr' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[23], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mmenr\u001b[49m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mDEM_topics_otsu_All\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mDEM_results(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTopic4\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'menr' is not defined"
     ]
    }
   ],
   "source": [
    "menr['DEM_topics_otsu_All'].DEM_results('Topic4')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now have completed all the steps necessary for starting the SCENIC+ analysis ðŸ˜….\n",
    "\n",
    "In particalular, we have\n",
    "\n",
    "1. preprocessed the scRNA-seq side of the data, selecting high quality cells and annotation these cells.\n",
    "2. preprocessed the scATAC-seq side of the data, selecting high quality cells, performing topic modeling and identifying candidate enhacer regions.\n",
    "3. looked for enriched motifs in candidate enhancer regions.\n",
    "\n",
    "In the next section we will combine all these analysis and run SCENIC+\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  },
  "vscode": {
   "interpreter": {
    "hash": "eb4bb617a32919d4a1b52f44f72e24e3739c7442a4f58a7bc491ec635ac76433"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
