#This was the command for the GEMOMA pipeline
/opt/jdk1.8.0_202/bin/java -Xmx120G -jar /opt/gemoma/GeMoMa-1.9.jar CLI ERE \ m=$bams/DRR047248-Ordenado.bam m=$bams/DRR047249-Ordenado.bam m=$bams/DRR047250-Ordenado.bam m=$bams/DRR047251-Ordenado.bam m=$bams/DRR047252-Ordenado.bam m=$bams/DRR047253-Ordenado.bam m=$bams/DRR047254-Ordenado.bam
/opt/jdk1.8.0_202/bin/java -Xmx120G -jar /opt/gemoma/GeMoMa-1.9.jar CLI CheckIntrons t=/home/phylobrain/bioinfo/gemoma/ref_genome.fa i=/work/rsenovilla/gemoma/ERE_MT/introns.gff
/opt/jdk1.8.0_202/bin/java -Xmx120G -jar /opt/gemoma/GeMoMa-1.9.jar CLI DenoiseIntrons i=/work/rsenovilla/gemoma/introns.gff coverage_unstranded=/work/rsenovilla/gemoma/coverage.bedgraph
nice -5 java -Xmx800G -jar /work/rsenovilla/gemoma-1.9/GeMoMa-1.9.jar CLI GeMoMaPipeline threads=70 >tblastn=false t=/work/rsenovilla/gemoma/ref_genome/ref_genome.fa r=EXTRACTED introns=/work/rsenovilla/gemoma/bams_mt/denoised_introns.gff coverage_unstranded=/work/rsenovilla/gemoma/bams_mt/coverage.bedgraph DenoiseIntrons.m=50000 GeMoMa.m=50000 GeMoMa.Score=ReAlign                                        s=own i=species1_name a=species1_annot g=species1_fasta #repeated for all species included
